src/redux/actions/authActions.js

```diff
@@ -0,0 +1,171 @@
+export const LOGOUT_REQUEST = 'LOGOUT_REQUEST';
+export const LOGOUT_FAILURE = 'LOGOUT_FAILURE';
+export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';
+export const REQUEST_USER_INFO = 'REQUEST_USER_INFO';
+export const RECEIVE_USER_INFO = 'RECEIVE_USER_INFO';
+
+function requestSignup() {
+  return {
+    type: SIGNUP_REQUEST,
+    receivedAt: Date.now()
+  };
+}
+
+function signupSuccess(user) {
+  return {
+    type: SIGNUP_SUCCESS,
+    user: user,
+    receivedAt: Date.now(),
+    isAuthenticated: true
+  };
+}
+
+function signupFailure() {
+  return {
+    type: SIGNUP_FAILURE,
+    receivedAt: Date.now()
+  };
+}
+
+function requestLogin() {
+  return {
+    type: LOGIN_REQUEST,
+    receivedAt: Date.now()
+  };
+}
+
+function loginSuccess() {
+  return {
+    type: LOGIN_SUCCESS,
+    receivedAt: Date.now(),
+    isAuthenticated: true
+  };
+}
+
+function loginFailure() {
+  return {
+    type: LOGIN_FAILURE,
+    receivedAt: Date.now()
+  };
+}
+
+function requestLogout() {
+  return {
+    type: LOGOUT_REQUEST,
+    receivedAt: Date.now()
+  };
+}
+
+function logoutSuccess() {
+  return {
+    type: LOGOUT_SUCCESS,
+    user: null,
+    receivedAt: Date.now(),
+    isAuthenticated: false
+  };
+}
+
+function logoutFailure() {
+  return {
+    type: LOGOUT_FAILURE,
+    receivedAt: Date.now()
+  };
+}
+
+function requestUserInfo() {
+  return {
+    type: REQUEST_USER_INFO,
+    receivedAt: Date.now()
+  };
+}
+
+function receiveUserInfo(user) {
+  return {
+    type: RECEIVE_USER_INFO,
+    user: user,
+    receivedAt: Date.now()
+  };
+}
+
+// TODO should fetching token be handled with a separate success/failure action?
+function fetchUserInfo() {
+  return dispatch => {
+    dispatch(requestUserInfo());
+    return fetch('http://localhost:4201/user/getinfo', {
+      headers: {
+        'Content-Type': 'application/json',
+        'x-access-token': localStorage.getItem('token')
+      },
+      method: 'get'
+    })
+      .then(response => response.json())
+      .then(json => dispatch(receiveUserInfo(json)));
+  };
+}
+
+/** Exported functions **/
+export function signup(user) {
+  return dispatch => {
+    dispatch(requestSignup(user)); // Signup request process has begun...
+    return fetch('http://localhost:4201/signup', {
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      method: 'post',
+      body: JSON.stringify(user)
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.success === false) dispatch(signupFailure());
+        else {
+          localStorage.setItem('token', json.token);
+          dispatch(signupSuccess(json.user));
+        }
+      });
+  };
+}
+
+export function login(user) {
+  return dispatch => {
+    dispatch(requestLogin(user)); // login request process has begun...
+    return fetch('http://localhost:4201/login', {
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      method: 'post',
+      body: JSON.stringify(user)
```

Comment:
I'm still not following what user is. In this case, the body is supposed to contain the following:

```
{
	"email" : "{{newemail}}",
	"password" : "{{password}}"
}
```

This is different from the last time user was used...

Reply:
The user parameter in signup is an object that is _expected_ to contain the keys email, password, f_name, and l_name - this is also true for login, minus f_name and l_name. Both our signup and login actions accept an undefined (damn JavaScript) object that should contain their required keys needed for their operation. We could make this more verbose by changing their signatures to

`export function signup(email, password, f_name, l_name)`
 
&

`export function login(email, password)`

which we would then just stringify into the body like

`body: JSON.stringify(email, password, ... etc)`

I'm undecided on which pattern I like more. Working with loosely typed objects is hurting my head a bit...

Reply:
I'm in support of keeping "user" but maybe add some documentation for this?

Reply:
That's a good idea. We could definitely include a parameter description in the method's documentation (which is a TODO task, I haven't included any documentation in the pull request yet - rip).

Reply:
Living for the fact that there are comments now. Beautiful.